### PR TITLE
BUG: Histogramdd breaks on big arrays in Windows

### LINKED ISSUE #22288: BUG: Histogramdd breaks on big arrays in Windows
### Describe the issue:

numpy.histogramdd fails with large arrays on Windows 10 64-bit. This is likely due to a bizarre behavior on Windows64 where the default integer dtype returns as int32 causing numpy.prod to overflow

### Reproduce the code example:

```python
>>> import numpy as np
>>> sample = np.zeros([100000000, 3])
>>> xbins = 400
>>> ybins = 400
>>> zbins = np.arange(16000)
>>> hist = np.histogramdd(sample=sample, bins=(xbins, ybins, zbins))
```

### Error message:

```shell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in histogramdd
  File "C:\Users\stephanies\AppData\Local\Continuum\miniconda3\envs\numpy_only\lib\site-packages\numpy\lib\histograms.py", line 1095, in histogramdd
    hist = np.bincount(xy, weights, minlength=nbin.prod())
  File "<__array_function__ internals>", line 180, in bincount
ValueError: 'minlength' must not be negative
```

### NumPy/Python version information:

1.23.3 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:29:51) [MSC v.1929 64 bit (AMD64)]

### Context for the issue:

This is a platform issue preventing the generation of large histograms. The same code above works fine on a linux machine

>>> import numpy as np
>>> sample = np.zeros([100000000, 3])
>>> xbins = 400
>>> ybins = 400
>>> zbins = np.arange(16000)
>>> hist = np.histogramdd(sample=sample, bins=(xbins, ybins, zbins))
>>> hist[0].shape
(400, 400, 15999)

### PR DESCRIPTION
**Pull request by navpreetnp7**
Resolved the issue with line change from int to np.intp in numpy/numpy/lib/histograms.py
New test function was added called test_big_arrays on numpy/lib/tests/test_histograms.py

Resolved #22288 
Pydata Sprint NYC 2022

### CODE REVIEW

File: numpy/lib/histograms.py [hunk 1]
```diff
@@ -970,7 +970,8 @@ def histogramdd(sample, bins=10, range=None, density=None, weights=None):
         sample = np.atleast_2d(sample).T
         N, D = sample.shape
 
-    nbin = np.empty(D, int)
+    #nbin = np.empty(D, int)
```
Comment (ganesh-k13) [inline numpy/lib/histograms.py:973]: Can you remove this line

Comment (navpreetnp7) [reply numpy/lib/histograms.py]: Yes, and committed.

File: numpy/lib/tests/test_histograms.py [hunk 2]
```diff
@@ -397,6 +397,14 @@ def test_histogram_bin_edges(self):
         edges = histogram_bin_edges(arr, bins='auto', range=(0, 1))
         assert_array_equal(edges, e)
 
+    def test_big_arrays(self):
```
Comment (seberg) [inline numpy/lib/tests/test_histograms.py:400]: Thanks.  We need to mark this for large memory use, have a look for `@requires_memory`.  I suspect it should also be marked with `@pytest.mark.slow` to skip on most runs.

**Comment by seberg** [comment]
There are two binary files that got accidentally commit, please delete them.

**Comment by seberg** [comment]
@navpreetnp7 would be great to finish this up, since branching will be in a week or two probably.  Could you add that `@requires_memory` and `@pytest.mark.slow` mark to the test?

CI/tests are currently failing because the test requires a huge amount of memory to run successfully.

**Comment by navpreetnp7** [comment]
Yeah I just finished it up. Sorry for the delay

File: numpy/lib/tests/test_histograms.py [hunk 3]
```diff
@@ -397,6 +398,16 @@ def test_histogram_bin_edges(self):
         edges = histogram_bin_edges(arr, bins='auto', range=(0, 1))
         assert_array_equal(edges, e)
 
+    @requires_memory(free_bytes=100000000 * 3)
```
Comment (seberg) [inline numpy/lib/tests/test_histograms.py:401]: Thanks.  Lets refine this a little bit.  If you check `sample.nbytes`, you will see that it is 8 times larger.   Also the result array (with the current choices) has a significant size.  You could just reduce that size by a lot, even then it probably makes sense to round up a little bit.
(This is a sanity check that we should be good after all.  The decorator also catches `MemoryError` so we should be fine, but I would just round up generously.)

Comment (navpreetnp7) [reply numpy/lib/tests/test_histograms.py]: Sure. That makes sense. I made it 100 times larger

Comment (seberg) [reply numpy/lib/tests/test_histograms.py]: OK, if I do the math, this should be about double of what is needed, and 10 GiB are not forbidding these days, so this should be fine.

**Comment by seberg** [comment]
Thanks @navpreetnp7.

**Pull request merged**